src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -55,69 +60,190 @@ protected override string CreateRequestId()
+            // ":scheme" is not restricted to "http" and "https" schemed URIs.  A
+            // proxy or gateway can translate requests for non - HTTP schemes,
+            // enabling the use of HTTP to interact with non - HTTP services.
+
+            // - That said, we shouldn't allow arbitrary values or use them to populate Request.Scheme, right?
+            // - For now we'll restrict it to http/s and require it match the transport.
+            // - We'll need to find some concrete scenarios to warrant unblocking this.
             if (!string.Equals(RequestHeaders[HeaderNames.Scheme], Scheme, StringComparison.OrdinalIgnoreCase))
             {
-                BadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestLine);
+                ResetAndAbort(new ConnectionAbortedException(
+                    CoreStrings.FormatHttp2StreamErrorSchemeMismatch(RequestHeaders[HeaderNames.Scheme], Scheme)), Http2ErrorCode.PROTOCOL_ERROR);
+                return false;
             }
 
+            // :path (and query) - Required
+            // Must start with / except may be * for OPTIONS
             var path = RequestHeaders[HeaderNames.Path].ToString();
-            var queryIndex = path.IndexOf('?');
+            RawTarget = path;
+
+            // OPTIONS - https://tools.ietf.org/html/rfc7540#section-8.1.2.3
+            // This pseudo-header field MUST NOT be empty for "http" or "https"
+            // URIs; "http" or "https" URIs that do not contain a path component
+            // MUST include a value of '/'.  The exception to this rule is an
+            // OPTIONS request for an "http" or "https" URI that does not include
+            // a path component; these MUST include a ":path" pseudo-header field
+            // with a value of '*'.
+            if (Method == HttpMethod.Options && path.Length == 1 && path[0] == '*')
+            {
+                // * is stored in RawTarget only since HttpRequest expects Path to be empty or start with a /.
+                Path = string.Empty;
+                QueryString = string.Empty;
+                return true;
+            }
 
-            Path = queryIndex == -1 ? path : path.Substring(0, queryIndex);
+            var queryIndex = path.IndexOf('?');
             QueryString = queryIndex == -1 ? string.Empty : path.Substring(queryIndex);
-            RawTarget = path;
 
-            // https://tools.ietf.org/html/rfc7230#section-5.4
-            // A server MUST respond with a 400 (Bad Request) status code to any
-            // HTTP/1.1 request message that lacks a Host header field and to any
-            // request message that contains more than one Host header field or a
-            // Host header field with an invalid field-value.
+            var pathSegment = queryIndex == -1 ? path.AsSpan() : path.AsSpan(0, queryIndex);
+
+            return TryValidatePath(pathSegment);
+        }
+
+        private bool TryValidateMethod()
```

Comment:
Curious why the `TryValidat*` naming. Might be more intuitive as `*IsValid`

Reply:
Try was more traditionally used for methods and Is for properties (e.g. persistent state), but otherwise there's no reason one way or the other.

Reply:
Actually, the big difference is that these methods have side effects, they ResetAndAbort if it fails. Not planning to rename.